system/include/SSE/emmintrin.h

```diff
@@ -0,0 +1,1621 @@
+_mm_cmpnlt_pd(__m128d __a, __m128d __b)
+{
+  return (__m128d)wasm_v128_not((v128_t)_mm_cmplt_pd(__a, __b));
+}
+
+static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
+_mm_cmpnle_pd(__m128d __a, __m128d __b)
+{
+  return (__m128d)wasm_v128_not((v128_t)_mm_cmple_pd(__a, __b));
+}
+
+static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
+_mm_cmpngt_pd(__m128d __a, __m128d __b)
+{
+  return (__m128d)wasm_v128_not((v128_t)_mm_cmpgt_pd(__a, __b));
+}
+
+static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
+_mm_cmpnge_pd(__m128d __a, __m128d __b)
+{
+  return (__m128d)wasm_v128_not((v128_t)_mm_cmpge_pd(__a, __b));
+}
+
+static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
+_mm_cmpeq_sd(__m128d __a, __m128d __b)
+{
+  return _mm_move_sd(__a, _mm_cmpeq_pd(__a, __b));
+}
+
+static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
+_mm_cmplt_sd(__m128d __a, __m128d __b)
+{
+  return _mm_move_sd(__a, _mm_cmplt_pd(__a, __b));
+}
+
+static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
+_mm_cmple_sd(__m128d __a, __m128d __b)
+{
+  return _mm_move_sd(__a, _mm_cmple_pd(__a, __b));
+}
+
+static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
+_mm_cmpgt_sd(__m128d __a, __m128d __b)
+{
+  return _mm_move_sd(__a, _mm_cmpgt_pd(__a, __b));
+}
+
+static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
+_mm_cmpge_sd(__m128d __a, __m128d __b)
+{
+  return _mm_move_sd(__a, _mm_cmpge_pd(__a, __b));
+}
+
+static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
+_mm_cmpord_sd(__m128d __a, __m128d __b)
+{
+  return _mm_move_sd(__a, _mm_cmpord_pd(__a, __b));
+}
+
+static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
+_mm_cmpunord_sd(__m128d __a, __m128d __b)
+{
+  return _mm_move_sd(__a, _mm_cmpunord_pd(__a, __b));
+}
+
+static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
+_mm_cmpneq_sd(__m128d __a, __m128d __b)
+{
+  return _mm_move_sd(__a, _mm_cmpneq_pd(__a, __b));
+}
+
+static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
+_mm_cmpnlt_sd(__m128d __a, __m128d __b)
+{
+  return _mm_move_sd(__a, _mm_cmpnlt_pd(__a, __b));
+}
+
+static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
+_mm_cmpnle_sd(__m128d __a, __m128d __b)
+{
+  return _mm_move_sd(__a, _mm_cmpnle_pd(__a, __b));
+}
+
+static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
+_mm_cmpngt_sd(__m128d __a, __m128d __b)
+{
+  return _mm_move_sd(__a, _mm_cmpngt_pd(__a, __b));
+}
+
+static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
+_mm_cmpnge_sd(__m128d __a, __m128d __b)
+{
+  return _mm_move_sd(__a, _mm_cmpnge_pd(__a, __b));
+}
+
+static __inline__ int __attribute__((__always_inline__, __nodebug__))
+_mm_comieq_sd(__m128d __a, __m128d __b)
+{
+  return ((__f64x2)__a)[0] == ((__f64x2)__b)[0];
```

Comment:
It would be more consistent to the intrinsic functions for these comparisons as well rather than falling back to using clang vector extensions.

Reply:
Marked down a note to compare how the intrinsic form performs.

Reply:
I would be shocked if they weren't equivalent. The intrinsic function is implemented using the clang vector extension `==` anyways, not a separate builtin function.

Reply:
Trying out both versions, they do indeed both end up with the same code:
```c++
#include <emmintrin.h>
#include <emscripten/html5.h>

__m128d __attribute__((noinline)) foo()
{
	return _mm_set1_pd(emscripten_performance_now());
}

int main()
{
	__m128d m = foo();
	__m128d m2 = foo();
	return _mm_comieq_sd(m, m2);
}
```

```bash
em++ a.cpp -o a.html -msse2 -g2 -O3 -fno-inline
```

both variants give

```wast
 (func $main (param $0 i32) (param $1 i32) (result i32)
  (i32.and
   (i32x4.extract_lane 0
    (f64x2.eq
     (f64x2.splat
      (call $emscripten_performance_now)
     )
     (f64x2.splat
      (call $emscripten_performance_now)
     )
    )
   )
   (i32.const 1)
  )
 )
```

Though I am surprised that both forms turn into a `f64x2.eq`. This will record more work to the underlying implementation, the VM will need to use `vcmppd`  instead of `comisd` instruction - unless some fancy pattern matching is used by the VM.

Looking at https://www.agner.org/optimize/instruction_tables.pdf, on Ryzen a `vcmpps` can only be issued only in ports 0 and 1, whereas a `comisd` can be issued in ports 0, 1 and 2. On Skylake-X, Agner does not unfortunately have the micro-port issue table for `vcmpps`, but latency for `comisd` is 2 cycles, whereas for `vcmppd` it is three cycles.

As a sidenote, uncovering the `i32.and` by `i32.const 1` makes me sad. 🙄

Converted these over to the intrinsics. To keep git amount of git wrangling low, the commit will come in in the SSE 3 PR.